File: notebook/july_evaluation/Scenario3/code/temperature_fit.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit
import scipy

# Set default text interpreter
plt.rc('text', usetex=True)
plt.rc('font', size=16) 

bl = '#0072BD'
br = '#D95319'

# Load data
data = scipy.io.loadmat('data.mat')
xdata = np.arange(len(data['tempData2'][0][:-1]))
ydata = data['tempData2'][0][:-1]

# Fit sinusoidal model f(x) = A*sin(Bx - C) + D
def fun(x, A, B, C, D):
    return A * np.sin(B * x - C) + D

#     A    B      C   D
p0 = [4, 0.02, 0, 15]
params, _ = curve_fit(fun, xdata, ydata, p0)

print(params)

plt.figure()
plt.plot(data['flowDate2'][0][:-1], ydata, '.', markersize=20, color=br)
plt.plot(xdata, fun(xdata, *params), linewidth=4, color=bl)
plt.ylabel('Temperature ($^\circ$C)')
plt.tight_layout()
plt.show()

File: notebook/july_evaluation/Scenario3/data/temperature_fit.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This file fits a sinusoidal function to the wastewater temperature data.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

rng('default');

clear;clc
set(groot, 'defaultAxesTickLabelInterpreter','latex');
set(groot, 'defaultLegendInterpreter','latex');
set(0,'defaulttextInterpreter','latex','defaultAxesFontSize',16) 
format long

bl = '#0072BD';
br = '#D95319';


%% Load data
load('data.mat')

xdata = 1:length(tempData2(1:end-1));
ydata = tempData2(1:end-1)';

% fit sinusoidal model f(x) = A*sin(Bx - C) + D
fun = @(p,xdata)(p(1)*sin(p(2)*xdata - p(3)) + p(4));

%     A    B      C   D
p0 = [4    0.02   0   15];
[params,sse] = lsqcurvefit(fun,p0,xdata,ydata);

disp(params)

figure()
    plot(flowDate2(1:end-1),ydata,'.','MarkerSize',20,'Color',br); hold on
    plot(xdata,fun(params,xdata),'LineWidth',4,'Color',bl); 
    xlim tight
    ylabel('Temperature ($^\circ$C)')
